### DISCUSSION TITLE
PicoGK in headless mode

### DISCUSSION THREAD

**Discussion by LinKayser** [leap71/PicoGK#30] [General]
As a result of many requests, PicoGK can now be run in headless mode, starting in 1.6. The caveat: You cannot access the viewer (obviously) and also not the log file. The idea is to have PicoGK run without any system interaction, except what you initiate yourself.

Here's an example:

```c#
try
{
    using (PicoGK.Library oLibrary = new(0.1f))
    {
        PicoGK.Voxels vox = new(PicoGK.Utils.mshCreateCube());
        vox.mshAsMesh().SaveToStlFile(Path.Combine(PicoGK.Utils.strDocumentsFolder(), "PicoGK_XS.stl"));
    }

    using (PicoGK.Library oLibrary = new(0.2f))
    {
        PicoGK.Voxels vox = new(PicoGK.Utils.mshCreateCube());
        vox.mshAsMesh().SaveToStlFile(Path.Combine(PicoGK.Utils.strDocumentsFolder(), "PicoGK_XS2.stl"));
    }

    using (PicoGK.Library oLibrary = new(0.3f))
    {
        PicoGK.Voxels vox = new(PicoGK.Utils.mshCreateCube());
        vox.mshAsMesh().SaveToStlFile(Path.Combine(PicoGK.Utils.strDocumentsFolder(), "PicoGK_XS3.stl"));
    }
}

catch (Exception e)
{
    Console.Write(e.ToString());
}

```

This showcases how to run the PicoGK library with several different voxel sizes and do something useful in between. Note the `using` syntax, which ensures proper cleanup, after you are done with your task.

This greatly simplifies running parameter sweeps, especially with different voxel sizes, etc.

It requires PicoGK 1.6.

Best,
Lin

**Comment by rtb21** [comment]
Thanks Lin, this is great!

**Comment by Bubbu17** [comment]
Hello Lin,

just a short question, i tried the code sample within my projekt but i
couldnt get it to work.
[image: image.png]
can see what the problem could be?
PikoGK Library available within the project?

best regards Torsten

**Comment by Bubbu17** [comment]
hold on maybe the version 1.6 is it...

**Comment by LinKayser** [comment]
You need to install the 1.6 runtime. The image isn't showing up for me, so I can't see the issue.

https://github.com/leap71/PicoGK/releases/tag/PicoGK-v1.6.0

Best,
Lin

**Comment by michaelp91-dev** [comment]
Hi Lin,

I am attempting to run PicoGK headless on a linux environment (I used Corrie's instructions to install it). I have succeeded in running PicoGK with the viewer, but when I try the code that is provided above so I can run it in headless mode I get the following two errors:

Program.cs(3,38): error CS1729: 'Library" does not contain a constructor that takes 1 arguments
Program.cs(3,12): error CS1674: 'Library': type used in a using statement must be implicitly convertible to 'System.IDisposable'

I am new to C# so I'm still learning, but if you or anyone has info that could help me resolve this issue, please let me know. Thank you!

**Reply by michaelp91-dev** [reply]
I have fixed this issue. The PicoGKRuntime i was running on was version 1.1. I recompiled it with the most recent version and it now works for me running PicoGK in headless mode.

**Reply by soigniert** [reply]
Hi michaelp91-dev,
Could please describe in more detail how you managed to compile the PicoGKRuntime? Which distribution and tool have you used?
As you can see [in this discussion](https://github.com/leap71/PicoGK/discussions/47), zwillikon and me struggled to compile it for Ubuntu. I'm working on my Win10 machine again, but would love to work under linux as well.
Best

**Reply by michaelp91-dev** [reply]
So all I was wanting to do was make a new .so file that was compiled from the PicoGKRuntime 1.7. I'll just detail everything I did and hopefully that helps. 

Details:
I work on the cloud mostly and this was all done on a Google Cloud Workstation running on Ubuntu 24.04

When I open my workstation I run this script to make sure I have all the dependencies needed. 


#!/bin/bash

set -e

sudo apt update
sudo apt install -y software-properties-common 
sudo apt-add-repository ppa:dotnet/backports -y

sudo apt update
sudo apt install -y \
    wget build-essential \
    libx11-dev \
    libxrandr-dev \
    libxinerama-dev \
    libxcursor-dev \
    libxi-dev \
    libboost-iostreams1.74-dev \
    libtbb-dev \
    libssl-dev \
    libzmq3-dev \
    libjemalloc-dev \
    libblosc-dev \
    doxygen \
    cmake-curses-gui \
    cmake-qt-gui \
    dotnet-sdk-7.0 \
    libwayland-dev \
    wayland-protocols \
    pkg-config \
    libxkbcommon-dev \
    clang

required_version="3.27.7"
binary_path="/usr/local/bin/cmake"

version_compare() {
    local req_ver=$1
    local inst_ver=$2
    if [[ "$(printf "%s\n" "$req_ver" "$inst_ver" | sort -V | head -n 1)" == "$req_ver" ]]; then
        return 0  
    else
        return 1   
    fi
}


installed_version=$(cmake --version | grep -oP "(?<=version )(\d+.\d+.\d+)")

version_compare "$required_version" "$installed_version"

if [[ $? -eq 0 ]]; then
    echo "Installed CMake version ($installed_version) is greater than or equal to $required_version.  Installation complete."
    exit 0
else
    echo "Installed CMake version ($installed_version) is less than $required_version. Adding CMake $required_version."
    wget https://github.com/Kitware/CMake/releases/download/v3.27.7/cmake-3.27.7-linux-x86_64.tar.gz
    tar -xzvf cmake-3.27.7-linux-x86_64.tar.gz
    sudo mv cmake-3.27.7-linux-x86_64 /opt/cmake-3.27.7
    if [ -e "$binary_path" ]; then
        echo -e "\033[36mTo Do: A version of cmake already exists at $binary_path. Cmake $required_version was installed at /opt/cmake.3.27.7. To use, add it to your PATH explicitly using the command: \n \n \
            export PATH=/opt/cmake.3.27.7/bin:\$PATH \n\033[0m"
    else
        sudo ln -s /opt/cmake-3.27.7/bin/cmake $binary_path
        echo "CMake $required_version symlinked to $binary_path from location in /opt."
    fi
    rm cmake-3.27.7-linux-x86_64.tar.gz
    echo "Installation complete."
fi
exit 0


After that's done then I run the following:
git clone --recursive https://github.com/leap71/PicoGKRuntime.git

After cloning, add this to the CMakeLists.txt file in PicoGKRuntime or it won't compile. I'm not sure if the order matters but I added it at line 91:
#enable position independent code
set(CMAKE_POSITION_INDEPENDENT_CODE ON)

Then I run these commands to compile:
cd PicoGKRuntime
mkdir build
cd build
cmake ..
make

After it compiles, the folder PicoGKRuntime/build/Lib will contain picogk.so and I can will copy that with the follow command:
sudo cp PicoGKRuntime/build/lib/*.so /usr/local/lib

Then in PicoGK, I open PicoGK__Config.cs and change this line:
public const string strPicoGKLib = "picogk.1.7";

To this:
public const string strPicoGKLib = "/usr/local/lib/picogk.so";




That's all. If that's a little hard to read, I apologize. Let me know if I can help further.

**Reply by joshmelnykcroft** [reply]
Thank you for this explanation! Based on this comment and a little tweaking ( not as much as you might think!) I got picoGK to run on my phone 🤣
![Screenshot_20250703_173829_TermuxX11.jpg](https://github.com/user-attachments/assets/b642590a-5756-44b9-b7e5-6baf03e17411)

**Reply by michaelp91-dev** [reply]
That's awesome! Are you using a terminal emulator to do that?

**Reply by joshmelnykcroft** [reply]
Exactly - termux with proot-distro for Ubuntu so I can get dotnet running, and a x-11 server so I can see the viewer.

**Reply by michaelp91-dev** [reply]
Thank you! I'm gonna try that as my hope is to create an STL file with PicoGK and then use FluidX3D (which works on Termux)  to run a Lattice-Boltzmann simulation on it.

**Reply by joshmelnykcroft** [reply]
I just looked at fluidx3d, looks very interesting! I wonder if there's an opportunity in the future to get a direct voxel import.
Let us know how that goes as CFD is a next step for me, very curious to see if you get that working!

**Comment by michaelp91-dev** [comment]
I wanted to share this if anyone else is trying to run PicoGK in headless mode. I was able to create a way to run PicoGK headless on Google Colab. I put it here in a repo: 

https://github.com/michaelp91-dev/PicoGK-Colab

I hope I'm not overstepping boundaries. But I wanted to share this if other people were wanting to do something this root. Obviously, with this, it gives you access to creating a STL file and then using within python, which is exactly what I plan to do with it. I need to run physics based simulations on what I design, and the physics engine I use can take in an STL file. I hope this is of use to others out there!

**Reply by rawwerks** [reply]
super helpful! thank you so much for sharing!
